core-blocks/archives/block.js

```diff
@@ -0,0 +1,82 @@
+/**
+ * WordPress dependencies
+ */
+import { Component, Fragment } from '@wordpress/element';
+import {
+	PanelBody,
+	ServerSideRender,
+	ToggleControl,
+} from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import {
+	InspectorControls,
+	BlockAlignmentToolbar,
+	BlockControls,
+} from '@wordpress/editor';
+
+class ArchivesBlock extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.toggleShowPostCounts = this.toggleShowPostCounts.bind( this );
+		this.toggleDisplayAsDropdown = this.toggleDisplayAsDropdown.bind( this );
+	}
+
+	toggleShowPostCounts() {
+		const { attributes, setAttributes } = this.props;
+		const { showPostCounts } = attributes;
+
+		setAttributes( { showPostCounts: ! showPostCounts } );
+	}
+
+	toggleDisplayAsDropdown() {
+		const { attributes, setAttributes } = this.props;
+		const { displayAsDropdown } = attributes;
+
+		setAttributes( { displayAsDropdown: ! displayAsDropdown } );
+	}
+
+	render() {
+		const { attributes, isSelected, setAttributes } = this.props;
+		const { align, showPostCounts, displayAsDropdown } = attributes;
+
+		const inspectorControls = isSelected && (
+			<InspectorControls key="inspector">
+				<PanelBody title={ __( 'Archives Settings' ) }>
+					<ToggleControl
+						label={ __( 'Show post counts' ) }
+						checked={ showPostCounts }
+						onChange={ this.toggleShowPostCounts }
+					/>
+					<ToggleControl
+						label={ __( 'Display as dropdown' ) }
+						checked={ displayAsDropdown }
+						onChange={ this.toggleDisplayAsDropdown }
+					/>
+				</PanelBody>
+			</InspectorControls>
+		);
+
+		return (
+			<Fragment>
+				{ inspectorControls }
+				<BlockControls key="controls">
+					<BlockAlignmentToolbar
+						value={ align }
+						onChange={ ( nextAlign ) => {
+							setAttributes( { align: nextAlign } );
+						} }
+						controls={ [ 'left', 'center', 'right', 'full' ] }
+					/>
+				</BlockControls>
+				<ServerSideRender key="archives" block="core/archives" attributes={ this.props.attributes } />
+			</Fragment>
+		);
+	}
+}
```

Comment:
Considering the life cycle of the ArchivesBlock component, and the fact that the block proper isn't very interactive, I'm not sure we need a class-based component and the ahead-of-time binding of the `toggle*` methods. In other words, `edit` could just be a functional component.

Reply:
@mcsf Curious if that's a matter of preference or would there be benefits for using `edit` as functional component instead?

Reply:
It's sort of been a tendency throughout Gutenberg to avoid what is, in a way, a performance trade-off that prioritizes performance of repeated rendering, at the (slight) cost of init-time bindings and increased memory footprint (which may add up with every little component) and at the (arguably) higher cost of code complexity for newcomers. For a component like ArchivesBlock, I'm tempted to say we're better off going with something simpler (i.e. functional component) even if it means binding two event handlers when rendering.

More on the subject of the cost of render-time inline functions: https://cdb.reacttraining.com/react-inline-functions-and-performance-bdff784f5578